Convert "shipment tracking providers" screen to MVVM

hichamboushaba opened the pull request:
The third and last part of #2892, it converts the screen `AddOrderTrackingProviderListFragment` to MVVM:

### Changes:
1. Converts AddOrderTrackingProviderListFragment to a regular fragment.
2. Update navigation to use navigation component.
3. Add an OrderShipmentProvidersRepository.
4. Converts the architecture to MVVM. 
5. Fixes an issue from my two previous PRs: cleaning up the repository to avoid leaking the eventbus subscription 1a82d753cb7ab4d5f4a7b2034c690137ee899f89

### Testing:
Test different scenarios of the AddOrderTrackingProviderListFragment.

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

hichamboushaba committed:
Add an app model and repository for shipment providers

hichamboushaba committed:
Add the dialog to the nav_graph

hichamboushaba committed:
Use nav component for going from tracking form to the providers list

hichamboushaba committed:
Finish the implementation of the ViewModel and wire it to the fragment

hichamboushaba committed:
Add tests

hichamboushaba committed:
Remove MVP classes

hichamboushaba committed:
Call the repository's onCleanup when the viewmodel is cleared

hichamboushaba committed:
Fix ktlint issues

0nko commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/tracking/AddOrderTrackingProviderListViewModel.kt:
```diff
@@ -0,0 +1,126 @@
+) : ScopedViewModel(savedState, dispatchers) {
+    private val navArgs: AddOrderTrackingProviderListFragmentArgs by savedState.navArgs()
+
+    val TrackingProviderListViewStateData = LiveDataDelegate(
```
Nit-pick: Normally we use lower-case letters for properties. I checked that the upper-case is used if it's a reference to an `object`.

hichamboushaba replied:
Good catch, autocomplete is useful, but sometimes it does thing like this 😄, I'll fix it 👍

0nko commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/tracking/AddOrderTrackingProviderListFragment.kt:
```diff
@@ -149,76 +78,67 @@ class AddOrderTrackingProviderListFragment : DialogFragment(), AddOrderTrackingP
-            return StringUtils.getCountryByCountryCode(it, presenter.loadStoreCountryFromDb())
+    private fun setupObservers(binding: DialogOrderTrackingProviderListBinding) {
+        viewModel.TrackingProviderListViewStateData.observe(viewLifecycleOwner) { old, new ->
+            new.providersList.takeIfNotEqualTo(old?.providersList) {
```
The `query` observer is missing, so the search doesn't work.

hichamboushaba replied:
I can't reproduce the issue of search, it's working for me, but I'll try to check it more, now that I think about it, having the same menu inflated here and in OrdersListFragment may cause an issue, as `onCreateOptionsMenu` will be called for both fragments (the same thing that caused the crash we just fixed)

But I don't think that the `query` has to be observed here, its value is just to restore the query text, which is done in `onCreateOptionsMenu`, and the filtering and propagating the filtered list is done here: https://github.com/woocommerce/woocommerce-android/blob/3ca9e474b7fe330981595059d1dcc2c330213fe0/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/tracking/AddOrderTrackingProviderListViewModel.kt#L78-L90

hichamboushaba replied:
@0nko I was able to confirm the issue I described above after screen rotation, I added the commit 157c6b848e06c58cb4237f549ac8b4fba4d30f4d to use a separate menu with a separate id for the search view to fix it.

0nko reviewed (changes requested):
The changes look good 👍, but the search doesn't work (more details in the code).

hichamboushaba committed:
Use a separate menu with separate id for the search view

as OrderListFragment use's the menu R.menu.search, and that it's active as a top level fragment, having the same id for the searchView causes its onPrepareOptionsMenu to unassign the search listeners after screen rotation.

0nko commented:
@hichamboushaba The search works now 👍. Just `TrackingProviderListViewStateData` needs to be changed and we can merge it.

hichamboushaba committed:
Use a lower case for the viewstate field to respect the conventions

hichamboushaba commented:
> @hichamboushaba The search works now 👍. Just `TrackingProviderListViewStateData` needs to be changed and we can merge it.

@0nko done eea1556, I forgot about it 🤦

0nko reviewed (approved):
Nice work! :shipit:
